src/operations/invalidate.ts

```diff
@@ -0,0 +1,62 @@
+import { forEachFieldNode } from './shared';
+import { Store, initStoreState, clearStoreState } from '../store';
+import { OperationRequest, Variables, Fragments, SelectionSet } from '../types';
+import {
+  getMainOperation,
+  normalizeVariables,
+  getFragments,
+  getSelectionSet,
+  getName,
+  getFieldArguments,
+} from '../ast';
+import { joinKeys, keyOfField } from '../helpers';
+
+interface Context {
+  store: Store;
+  variables: Variables;
+  fragments: Fragments;
+}
+
+export const invalidate = (store: Store, request: OperationRequest) => {
+  initStoreState(0);
+  const operation = getMainOperation(request.query);
+
+  const ctx: Context = {
+    variables: normalizeVariables(operation, request.variables),
+    fragments: getFragments(request.query),
+    store,
+  };
+
+  invalidateSelection(ctx, 'Query', getSelectionSet(operation));
+
+  clearStoreState();
+};
+
+export const invalidateSelection = (
+  ctx: Context,
+  entityKey: string,
+  select: SelectionSet
+) => {
+  const { store, variables } = ctx;
+  const typename = store.getField(entityKey, '__typename');
+  if (typeof typename !== 'string') return null;
+
+  forEachFieldNode(typename, entityKey, select, ctx, node => {
+    const fieldName = getName(node);
+    const fieldArgs = getFieldArguments(node, variables);
+    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, fieldArgs));
+
+    if (node.selectionSet === undefined) {
+      store.removeRecord(fieldKey);
+    } else {
+      const fieldSelect = getSelectionSet(node);
+      const link = store.getLink(fieldKey);
+      store.removeLink(fieldKey);
```

Comment:
There's a special case here where this may still be a record I think 🤔

Reply:
Looking this up, so basically when `fieldValue` is not undefined we want to delete that as well. It may be beneficial to actually retrieve the `record` when the `link` is `undefined` and check whether it needs to be removed